File: medication_timer.go
// medication_timer.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"time"
)

type Medication struct {
	ID           int       `json:"id"`
	Name         string    `json:"name"`
	Dosage       string    `json:"dosage"`
	IntervalHours float64  `json:"interval_hours"`
	LastTaken    time.Time `json:"last_taken"`
	History      []time.Time `json:"history"`
}

func (m *Medication) IsDue() bool {
	next := m.LastTaken.Add(time.Duration(m.IntervalHours * float64(time.Hour)))
	return time.Now().After(next) || time.Now().Equal(next)
}

func (m *Medication) NextDose() time.Time {
	return m.LastTaken.Add(time.Duration(m.IntervalHours * float64(time.Hour)))
}

func (m *Medication) TakeDose() {
	now := time.Now()
	m.LastTaken = now
	m.History = append(m.History, now)
}

type TimerApp struct {
	Medications []Medication `json:"medications"`
	File        string
}

func NewTimerApp(file string) *TimerApp {
	app := &TimerApp{File: file}
	app.load()
	return app
}

func (t *TimerApp) load() {
	data, err := os.ReadFile(t.File)
	if err != nil {
		return
	}
	json.Unmarshal(data, t)
}

func (t *TimerApp) save() {
	data, _ := json.MarshalIndent(t, "", "  ")
	os.WriteFile(t.File, data, 0644)
}

func (t *TimerApp) AddMedication(name, dosage string, interval float64) {
	maxID := 0
	for _, m := range t.Medications {
		if m.ID > maxID {
			maxID = m.ID
		}
	}
	med := Medication{
		ID:            maxID + 1,
		Name:          name,
		Dosage:        dosage,
		IntervalHours: interval,
		LastTaken:     time.Now(),
		History:       []time.Time{},
	}
	t.Medications = append(t.Medications, med)
	t.save()
	fmt.Printf("💊 Added medication: %s %s every %.1fh (ID: %d)\n", name, dosage, interval, med.ID)
}

func (t *TimerApp) List() {
	if len(t.Medications) == 0 {
		fmt.Println("No medications.")
		return
	}
	fmt.Println("\n📋 Medications:")
	for _, m := range t.Medications {
		due := "🔴 DUE"
		if !m.IsDue() {
			due = fmt.Sprintf("⏳ %02d:%02d", m.NextDose().Hour(), m.NextDose().Minute())
		}
		fmt.Printf("  [%d] %s %s – %s\n", m.ID, m.Name, m.Dosage, due)
	}
}

func (t *TimerApp) TakeDose(id int) {
	for i, m := range t.Medications {
		if m.ID == id {
			t.Medications[i].TakeDose()
			t.save()
			fmt.Printf("✅ Took %s %s at %s\n", m.Name, m.Dosage, time.Now().Format("15:04"))
			return
		}
	}
	fmt.Printf("❌ Medication with ID %d not found.\n", id)
}

func (t *TimerApp) CheckDue() {
	due := []Medication{}
	for _, m := range t.Medications {
		if m.IsDue() {
			due = append(due, m)
		}
	}
	if len(due) == 0 {
		fmt.Println("✅ No medications due right now.")
		return
	}
	fmt.Println("\n⏰ Due Medications:")
	for _, m := range due {
		overdue := int(time.Now().Sub(m.NextDose()).Minutes())
		fmt.Printf("  [%d] %s %s – overdue by %d min\n", m.ID, m.Name, m.Dosage, overdue)
	}
}

func (t *TimerApp) History() {
	if len(t.Medications) == 0 {
		fmt.Println("No medications.")
		return
	}
	fmt.Println("\n📜 History (last 5 doses per medication):")
	for _, m := range t.Medications {
		fmt.Printf("  %s:\n", m.Name)
		recent := m.History
		if len(recent) > 5 {
			recent = recent[len(recent)-5:]
		}
		if len(recent) == 0 {
			fmt.Println("    (no history)")
		} else {
			for _, ts := range recent {
				fmt.Printf("    - %s\n", ts.Format("2006-01-02 15:04"))
			}
		}
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: medication_timer <command> [options]")
		return
	}
	app := NewTimerApp("medications.json")
	cmd := os.Args[1]
	switch cmd {
	case "add":
		if len(os.Args) < 5 {
			fmt.Println("add <name> <dosage> <interval_hours>")
			return
		}
		name := os.Args[2]
		dosage := os.Args[3]
		interval, _ := strconv.ParseFloat(os.Args[4], 64)
		app.AddMedication(name, dosage, interval)
	case "list":
		app.List()
	case "take":
		if len(os.Args) < 3 {
			fmt.Println("take <id>")
			return
		}
		id, _ := strconv.Atoi(os.Args[2])
		app.TakeDose(id)
	case "check":
		app.CheckDue()
	case "history":
		app.History()
	default:
		fmt.Println("Unknown command.")
	}
}

File: medication_timer.php
# medication_timer.php
<?php
$dataFile = 'medications.json';

class Medication {
    public $id, $name, $dosage, $interval_hours, $last_taken, $history;

    function __construct($name, $dosage, $interval_hours, $id = 0, $last_taken = null, $history = []) {
        $this->id = $id;
        $this->name = $name;
        $this->dosage = $dosage;
        $this->interval_hours = $interval_hours;
        $this->last_taken = $last_taken ? new DateTime($last_taken) : new DateTime();
        $this->history = array_map(function($t) { return new DateTime($t); }, $history);
    }

    function isDue() {
        $next = clone $this->last_taken;
        $next->modify("+{$this->interval_hours} hours");
        return new DateTime() >= $next;
    }

    function nextDose() {
        $next = clone $this->last_taken;
        $next->modify("+{$this->interval_hours} hours");
        return $next;
    }

    function takeDose() {
        $now = new DateTime();
        $this->last_taken = $now;
        $this->history[] = $now;
    }

    function toArray() {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'dosage' => $this->dosage,
            'interval_hours' => $this->interval_hours,
            'last_taken' => $this->last_taken->format(DateTime::ISO8601),
            'history' => array_map(function($d) { return $d->format(DateTime::ISO8601); }, $this->history)
        ];
    }

    static function fromArray($data) {
        return new self($data['name'], $data['dosage'], $data['interval_hours'], $data['id'], $data['last_taken'], $data['history']);
    }
}

class TimerApp {
    private $medications = [];
    private $file;

    function __construct($file) {
        $this->file = $file;
        $this->load();
    }

    function load() {
        if (file_exists($this->file)) {
            $data = json_decode(file_get_contents($this->file), true);
            foreach ($data as $item) {
                $this->medications[] = Medication::fromArray($item);
            }
        }
    }

    function save() {
        $data = array_map(function($m) { return $m->toArray(); }, $this->medications);
        file_put_contents($this->file, json_encode($data, JSON_PRETTY_PRINT));
    }

    function addMedication($name, $dosage, $interval) {
        $maxId = 0;
        foreach ($this->medications as $m) if ($m->id > $maxId) $maxId = $m->id;
        $med = new Medication($name, $dosage, $interval, $maxId + 1);
        $this->medications[] = $med;
        $this->save();
        echo "💊 Added medication: $name $dosage every $interval" . "h (ID: {$med->id})\n";
    }

    function list() {
        if (empty($this->medications)) {
            echo "No medications.\n";
            return;
        }
        echo "\n📋 Medications:\n";
        foreach ($this->medications as $m) {
            $due = $m->isDue() ? "\033[31m🔴 DUE\033[0m" : "\033[32m⏳ " . $m->nextDose()->format('H:i') . "\033[0m";
            echo "  [{$m->id}] {$m->name} {$m->dosage} – $due\n";
        }
    }

    function takeDose($id) {
        foreach ($this->medications as $m) {
            if ($m->id == $id) {
                $m->takeDose();
                $this->save();
                echo "✅ Took {$m->name} {$m->dosage} at " . (new DateTime())->format('H:i') . "\n";
                return;
            }
        }
        echo "❌ Medication with ID $id not found.\n";
    }

    function checkDue() {
        $due = array_filter($this->medications, function($m) { return $m->isDue(); });
        if (empty($due)) {
            echo "✅ No medications due right now.\n";
            return;
        }
        echo "\n⏰ Due Medications:\n";
        foreach ($due as $m) {
            $overdue = (int)((time() - $m->nextDose()->getTimestamp()) / 60);
            echo "  [{$m->id}] {$m->name} {$m->dosage} – overdue by $overdue min\n";
        }
    }

    function history() {
        if (empty($this->medications)) {
            echo "No medications.\n";
            return;
        }
        echo "\n📜 History (last 5 doses per medication):\n";
        foreach ($this->medications as $m) {
            echo "  {$m->name}:\n";
            $recent = array_slice($m->history, -5);
            if (empty($recent)) {
                echo "    (no history)\n";
            } else {
                foreach ($recent as $dt) {
                    echo "    - " . $dt->format('Y-m-d H:i') . "\n";
                }
            }
        }
    }
}

if ($argc < 2) {
    die("Usage: php medication_timer.php <command> [options]\n");
}
$app = new TimerApp($dataFile);
$cmd = $argv[1];

switch ($cmd) {
    case 'add':
        if ($argc < 5) die("add <name> <dosage> <interval_hours>\n");
        $app->addMedication($argv[2], $argv[3], (float)$argv[4]);
        break;
    case 'list':
        $app->list();
        break;
    case 'take':
        if ($argc < 3) die("take <id>\n");
        $app->takeDose((int)$argv[2]);
        break;
    case 'check':
        $app->checkDue();
        break;
    case 'history':
        $app->history();
        break;
    default:
        echo "Unknown command.\n";
}
?>
